src/jit/gentree.cpp

```diff
@@ -9183,231 +9183,340 @@ GenTreePtr GenTree::GetChild(unsigned childNum)
-                case 1:
-                    return &m_node->AsArrOffs()->gtIndex;
-                case 2:
-                    return &m_node->AsArrOffs()->gtArrObj;
-                default:
-                    return nullptr;
-            }
+            case GT_ARR_OFFSET:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceArrOffset;
+                break;
+
+            case GT_DYN_BLK:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceDynBlk;
+                break;
+
+            case GT_STORE_DYN_BLK:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceStoreDynBlk;
+                break;
+
+            case GT_LEA:
+                if (m_node->AsAddrMode()->gtOp1 == nullptr)
+                {
+                    m_advance = &GenTreeUseEdgeIterator::AdvanceLea;
+                }
+                else
+                {
+                    m_advance = GetAdvanceFnForBinOp(m_node);
+                }
+                break;
+
+            case GT_PHI:
+                m_argList = m_node->AsUnOp()->gtOp1;
+                m_advance = &GenTreeUseEdgeIterator::AdvanceList;
+                break;
 
-        // Call, phi, and SIMD nodes are handled by MoveNext{Call,Phi,SIMD}UseEdge, repsectively.
-        case GT_CALL:
-        case GT_PHI:
 #ifdef FEATURE_SIMD
-        case GT_SIMD:
+            case GT_SIMD:
+                if (m_node->AsSIMD()->gtSIMDIntrinsicID == SIMDIntrinsicInitN)
+                {
+                    m_argList = node->AsSIMD()->gtOp1;
+                    m_advance = &GenTreeUseEdgeIterator::AdvanceList;
+                }
+                else
+                {
+                    m_advance = GetAdvanceFnForBinOp(m_node);
+                }
+                break;
 #endif
-            break;
 
-        case GT_ASG:
+            case GT_FIELD_LIST:
+                m_argList = node;
+                m_advance = &GenTreeUseEdgeIterator::AdvanceList;
+                break;
+
+            case GT_CALL:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceCall;
+                break;
+
+            default:
+                if (m_node->OperIsUnary())
+                {
+                    GenTree** edge = &m_node->AsUnOp()->gtOp1;
+                    if (*edge == nullptr)
+                    {
+                        m_state = -1;
+                    }
+                    else
+                    {
+                        m_edge = edge;
+                        m_advance = &GenTreeUseEdgeIterator::AdvanceUnary;
+                    }
+                    return;
+                }
+                else
+                {
+                    m_advance = GetAdvanceFnForBinOp(m_node);
+                }
+                break;
+        }
+
+        // Advance to the first operand.
+        ++(*this);
+    }
+}
+
+void GenTreeUseEdgeIterator::AdvanceCmpXchg()
+{
+    FinishSimpleAdvance([this]() -> GenTree** {
+        switch (m_state)
```

Comment:
I considered this but dismissed it out of hand because it requires memory to hold the table that in the general case (i.e. nodes with an unknown number of arguments) must be dynamically allocated and the table needs to be setup for each instance of the iterator.

That said, this might be viable for smallish nodes where the setup cost is low and the table can be cheaply preallocated in the iterator. Note that this would also benefit from a "next state" table (rather than just the "next edge" table) in order to remove branches, which would use a bit more memory.

Perhaps a hybrid solution is best: use static tables of pointers to data members and next states. As mentioned, though, I'll need to get some profile data to really know.